### PR TITLE
Add permission check for roles

### PR DESCRIPTION
**Pull request by Plerx2493**
# Summary
Adds a overload for `PermissonsFor` to allow checking permissions for a specific role

### CODE REVIEW

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 1]
```diff
@@ -1101,6 +1101,54 @@ public Permissions PermissionsFor(DiscordMember mbr)
...
+    {
+        if (this.IsThread)
+        {
+            return this.Parent.PermissionsFor(role);
+        }
+        
+        if (this.IsPrivate || this.Guild is null)
+        {
+            return Permissions.None;
+        }
```
Comment (akiraveliara) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:1119]: wouldn't we also want to check whether the role and channel belong to the same guild

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 2]
```diff
@@ -1101,6 +1101,54 @@ public Permissions PermissionsFor(DiscordMember mbr)
...
+        perms |= role.Permissions;
+        
+        // Administrator grants all permissions and cannot be overridden
+        if ((perms & Permissions.Administrator) == Permissions.Administrator)
+        {
+            return PermissionMethods.FULL_PERMS;
+        }
+        
+        // channel overrides for the role
+        DiscordOverwrite? roleOverwrites = this._permissionOverwrites.FirstOrDefault(xo => xo.Id == role.Id);
```
Comment (akiraveliara) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:1137]: what about channel overwrites for the @ everyone role? are those technically possible?

**Review by akiraveliara — CHANGES_REQUESTED** [review]
i'd like the check addressed, personally

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 3]
```diff
@@ -1101,6 +1101,70 @@ public Permissions PermissionsFor(DiscordMember mbr)
...
+        perms |= role.Permissions;
+        
+        // Administrator grants all permissions and cannot be overridden
+        if ((perms & Permissions.Administrator) == Permissions.Administrator)
+        {
+            return PermissionMethods.FULL_PERMS;
+        }
+        
+        // channel overrides for the role
+        DiscordOverwrite? everyoneRoleOverwrites = this._permissionOverwrites.FirstOrDefault(xo => xo.Id == role.Id);
```
Comment (akiraveliara) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:1142]: wrong ID

**Review by akiraveliara — CHANGES_REQUESTED** [review]

File: DSharpPlus/Entities/Channel/DiscordChannel.cs (hunk 3 above)
Comment (Plerx2493) [reply DSharpPlus/Entities/Channel/DiscordChannel.cs:1142]: ups - c&p

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 4]
```diff
@@ -1101,6 +1101,77 @@ public Permissions PermissionsFor(DiscordMember mbr)
...
+            // assign channel permission overwrites for the role (explicit deny)
+            perms &= ~everyoneRoleOverwrites.Denied;
+        
+            // assign channel permission overwrites for the role (explicit allow)
+            perms |= everyoneRoleOverwrites.Allowed;
+        }
+        
+        // channel overrides for the role
+        DiscordOverwrite? roleOverwrites = this._permissionOverwrites.FirstOrDefault(xo => xo.Id == role.Id);
+        if (roleOverwrites == null)
```
Comment (akiraveliara) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:1154]: nit: `is null`

**Review by akiraveliara — APPROVED** [review]

**Pull request merged**